cxx4opencl/address_spaces.txt

```diff
@@ -209,46 +292,61 @@ __kernel void foo() {
-  [&] () __private mutable {} (); // error: mutable specifier should precede address space.
+  [&] () __private mutable {} (); // error: mutable specifier should precede address
+                                  // space.
 }
-----------
-
+------------
 
 ==== Implicit special members
 
 The prototype for implicit special members (default, copy or move constructor,
-copy or move assignment, destructor) has the generic address space for an implicit
+copy or move assignment, destructor) has the default address space for an implicit
 object pointer and reference parameters (see also
-<<addrspace-member-function-qualifiers, _Member function qualifier_>>).
+<<addrspace-member-function-qualifiers, _Member function qualifier_>>). This default
+address space is generic if it is supported or private otherwise.
 
 [source,cpp]
-----------
+------------
 class C {
   // Has the following implicitly defined member functions.
 
-  // void C(); /* implicit 'this' parameter is a pointer to */
-               /* object in generic address space. */
+  // C(); /* implicit 'this' parameter is a pointer to */
+          /* object in generic address space if supported */
+          /* in private address space otherwise. */
 
-  // void C(const C & par); /* 'this'/'par' is a pointer/reference to */
-                            /*  object in generic address space. */
+  // C(const C & par); /* 'this'/'par' is a pointer/reference to */
+                       /*  object in generic address space */
+                       /*  if supported */
+                       /* in private address space otherwise. */
 
-  // void C(C && par); /* 'this'/'par' is a pointer/r-val reference to */
-                       /* object in generic address space. */
+  // C(C && par); /* 'this'/'par' is a pointer/r-val reference to */
+                  /* object in generic address space if supported */
+                  /* in private address space otherwise. */
 
-  // operator= C &(const C & par); /* 'this'/'par'/return value is */
+  // C & operator=(const C & par); /* 'this'/'par'/return value is */
                                    /* a pointer/reference/reference to */
-                                   /* object in generic address space. */
+                                   /* object in generic address space */
+                                   /* if supported */ 
+                                   /* in private address space otherwise. */
 
-  // operator= C &(C && par)'; /* 'this'/'par'/return value is */
+  // C & operator=(C && par)'; /* 'this'/'par'/return value is */
                                /* a pointer/r-val reference/reference to */
                                /* object in generic address space. */
+                               /* if supported */ 
+                               /* in private address space otherwise. */
 };
-----------
+------------
+
+`TODO: As an optimisation we can do an upgrade in this version and always infer
```

Comment:
Is this TODO meant to stay and be published?

Reply:
No I am going to remove this since we didn't have time to implement this in clang yet.